### ISSUE TITLE
opt.train_sample_n is 16?

### ISSUE THREAD

**Issue by homelifes** [ruotianluo/self-critical.pytorch#180]
Hi @ruotianluo. May i ask why you are using train_sample_n = 16 for gen_result and n=1 for greedy in the code below? In `transformer_sc.yml`, you dont define train_sample_n so it should follow the value in `opts` which is 16. Shouldn't this be >1 only when using new self-critical when we want to generate multiple samples?
```
          self.model.eval()
            with torch.no_grad():
                greedy_res, _ = self.model(fc_feats, att_feats, att_masks,
                    mode='sample',
                    opt={'sample_method': opt.sc_sample_method,
                         'beam_size': opt.sc_beam_size})
            self.model.train()
            gen_result, sample_logprobs = self.model(fc_feats, att_feats, att_masks,
                    opt={'sample_method':opt.train_sample_method,
                        'beam_size':opt.train_beam_size,
                        'sample_n': opt.train_sample_n},
                    mode='sample')
```

**Comment by ruotianluo** [comment]
Train sample n is defined in transformer.yml

**Comment by homelifes** [comment]
In that case, why is sampling 5x more than greedy? You repeated the tensor 5 times for sampling but one time for greedy. And then when you computed the cider score, in which
`greedy_res.shape[0] ==batch_size`  and `gen_result_size = batch_size * seq_per_img`
and then:
```
    for i in range(gen_result_size):
        res[i] = [array_to_str(gen_result[i])]
    for i in range(batch_size):
        res[gen_result_size + i] = [array_to_str(greedy_res[i])]
    gts = OrderedDict()
    for i in range(len(data_gts)):
        gts[i] = [array_to_str(data_gts[i][j]) for j in range(len(data_gts[i]))]

    res_ = [{'image_id':i, 'caption': res[i]} for i in range(len(res))]
    gts_ = {i: gts[i // seq_per_img] for i in range(gen_result_size)}
    gts_.update({i+gen_result_size: gts[i] for i in range(batch_size)})
```
So here the sampling is more than greedy. This wasn't in your previous version. May I know what is the purpose of changing it from the previous version and why is the greedy baseline not equal to the sample when subtracted from each other?

**Comment by ruotianluo** [comment]
The purpose of changing it is to reduce unnecessary computation. Since the model is set to eval mode, sample 5 greedy decoding is the same as sampling 1 greedy decoding.

**Comment by homelifes** [comment]
@ruotianluo do you mean that in sampling, we can have different outputs for the same image, but for greedy its only one output since we are taking the max. So we only perform greedy one time and duplicate later? Is that what you mean?

**Comment by ruotianluo** [comment]
yes.

**Comment by homelifes** [comment]
alright. Thanks a lot. I have one more question.
I realized that in gathering the logprobs:
```
           it, sampleLogprobs = self.sample_next_word(logprobs, sample_method, temperature)

            # stop when all finished
            if t == 0:
                unfinished = it > 0
            else:
                unfinished = unfinished * (it > 0)
            it = it * unfinished.type_as(it)
            seq[:,t] = it
            seqLogprobs[:,t] = logprobs
```
You are saving the `logprobs` and not the sampled ones `sampleLogprobs`. And then in  `RewardCriterion` you gather them according to `it`:  
`input = input.gather(2, seq.unsqueeze(2)).squeeze(2)`

May i also if that contributes to performance? Or is it just another way?

**Comment by ruotianluo** [comment]
No. It's because some other loss may need the full probability.
